internal/servers/controller/handlers/authmethods/oidc.go

```diff
@@ -150,94 +150,124 @@ func (s Service) authenticateOidc(ctx context.Context, req *pbs.AuthenticateRequ
 	}
 
-	resp := &pb.OidcAuthMethodAuthenticateStartResponse{
-		AuthUrl:  authUrl.String(),
-		TokenUrl: tokenUrl.String(),
-		TokenId:  tokenId,
+	respAttrs := &pb.OidcAuthMethodAuthenticateStartResponse{
+		AuthUrl: authUrl.String(),
+		TokenId: tokenId,
 	}
-
-	attrs, err := handlers.ProtoToStruct(resp)
-	if err != nil {
+	resp := &pbs.AuthenticateResponse{Command: req.GetCommand()}
+	if resp.Attributes, err = handlers.ProtoToStruct(respAttrs); err != nil {
 		return nil, errors.New(errors.Internal, op, "Error marshaling parameters.")
 	}
-
-	return &pbs.AuthenticateResponse{Command: req.GetCommand(), Attributes: attrs}, nil
+	return resp, nil
 }
 
-func (s Service) authenticateOidcCallback(ctx context.Context, req *pbs.AuthenticateRequest, authResults *auth.VerifyResults) (*pbs.AuthenticateResponse, error) {
+// authenticateOidcCallback behaves differently than other service methods.
+// Because of the way it this is called by the end user, it should only return
+// an error if we are unable to lookup the auth method or the request
+// parameters were invalid.  All other errors should be returned back through
+// the response as a finalRedirectUrl to an endpoint that can properly show the
+// error details.
+func (s Service) authenticateOidcCallback(ctx context.Context, req *pbs.AuthenticateRequest) (*pbs.AuthenticateResponse, error) {
 	const op = "authmethod_service.(Service).authenticateOidcCallback"
 	if req == nil {
 		return nil, errors.New(errors.InvalidParameter, op, "Nil request.")
 	}
-	if authResults == nil {
-		return nil, errors.New(errors.InvalidParameter, op, "Nil auth results.")
+
+	repo, err := s.oidcRepoFn()
+	if err != nil {
+		return nil, errors.Wrap(err, op)
+	}
+	am, err := repo.LookupAuthMethod(ctx, req.GetAuthMethodId())
+	if err != nil {
+		return nil, errors.Wrap(err, op)
+	}
+	if am == nil {
+		return nil, errors.New(errors.RecordNotFound, op, fmt.Sprintf("auth method %s not found", req.GetAuthMethodId()))
+	}
+	if am.GetApiUrl() == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "auth method doesn't have api url defined")
 	}
+	errResponse := errorResponseBuilder(am.GetApiUrl())
 
 	attrs := new(pb.OidcAuthMethodAuthenticateCallbackRequest)
 	// Note that this conversion has already happened in the validate call so we don't expect errors here.
 	if err := handlers.StructToProto(req.GetAttributes(), attrs, handlers.WithDiscardUnknownFields(true)); err != nil {
-		return nil, err
+		return errResponse(err), nil
 	}
 
-	finalRedirectUrl, err := oidc.Callback(
+	var finalRedirectUrl string
+	if attrs.GetError() != "" {
+		// TODO: Package the OIDC error into the redirectUrl
+		return errResponse(fmt.Errorf("Error: %q, Details: %q", attrs.GetError(), attrs.GetErrorDescription())), nil
+	}
+	finalRedirectUrl, err = oidc.Callback(
 		ctx,
 		s.oidcRepoFn,
 		oidc.IamRepoFactory(s.iamRepoFn),
 		s.atRepoFn,
-		req.GetAuthMethodId(),
+		am,
```

Comment:
why not just pass the am.PublicId?

Reply:
The auth method is needed in this calling method to redirect for errors in this method.  Instead of doing the lookup twice I passed it through.

Reply:
Just to follow up, relooking at the code the called method does not need the full auth method.  I must have confused my initial iteration on this fix, where the called method also generates the redirectUrl with what is actually there.  I'll fix this and re-request review.  Thanks!

Reply:
follow up to the follow up: The auth method is needed for looking up the provider, getting scope, state, max age, etc...  Leaving it as is.